### PR TITLE
Protect against infinite reboot loops

### LINKED ISSUE #81: Prevent an infinite reboot loop
To prevent a benchmarking machine rebooting infinitely:

  1. Store the number of reboots to date in the results file
  2. On starting krun, *after* the execution schedule has been build, add together the number of executions completed so far, with the number planned
  3. If the value from step 2. is less than the number of reboots so far, halt with an `INFO` message in the log

### PR DESCRIPTION
**Pull request by snim2**
This PR:

  * Adds a 'reboots' key to results files
  * Calculates the expected number of reboots from the configuration file
  * Halts Krun if the next reboot is greater than the number of expected reboots
  * Updates the tests
  * Adds a `--dump-reboots` option to the CLI
  * Documents the new layout of the results file and the `--dump-reboots` option.

Fixes #81

### CODE REVIEW

File: krun/scheduler.py [hunk 1]
```diff
@@ -341,6 +344,18 @@ def run(self):
...
     def _reboot(self):
+        self.nreboots += 1
+        debug("About to execute reboot: %g, expecting %g in total." %
+              (self.nreboots, self.expected_reboots))
+        # Dump the results file. This may already have been done, but we
+        # have changed self.nreboots, which needs to be written out.
+        util.dump_results(self.config_file, self.out_file, self.results,
+                  self.platform.audit, self.nreboots)
+        if self.nreboots > self.expected_reboots:
+            util.fatal(("About to execute reboot number: %g. " +
```
Comment (vext01) [inline krun/scheduler.py]: Can we be more clear that something has gone wrong?

**Comment by vext01** [comment]
Aside from that minor quibble, looks good.

If it's tested, I would be happy to merge.

**Comment by snim2** [comment]
OK, this last commit changes the error message to:

```
[2015-11-06 14:00:09: ERROR] HALTING now to prevent an infinite reboot loop: about to execute reboot number: 9. 8 jobs have been completed, 7 are left to go.
```

This has been tested (the easiest thing I found was to comment out the line `self.work_deque.remove(job_to_remove)` in the scheduler`)

**Comment by vext01** [comment]
Ah, sorry I wasn't clear. I think the message should include something to the effect of "this shouldn't happen".

**Comment by snim2** [comment]
OK, if you are happy with:

```
[2015-11-06 14:07:17: ERROR] HALTING now to prevent an infinite reboot loop, INVARIANT num_reboots <= num_jobs violated. Krun was about to execute reboot number: 9. 8 jobs have been completed, 7 are left to go.
```

I will squash (don't worry about the odd numbers, that was just my hack to create an infinite schedule).

**Comment by vext01** [comment]
Looks good. Go ahead and squash.

**Comment by snim2** [comment]
OK, squashed and brought up to date with master.

**Pull request merged**